lib/app/services/database/conversation_db_service.c.dart

```diff
@@ -0,0 +1,345 @@
+          mode: InsertMode.insertOrReplace,
+        );
+
+    if (eventMessage.kind == PrivateDirectMessageEntity.kind) {
+      await _updateConversationMessagesTable(eventMessage);
+    } else if (eventMessage.kind == PrivateMessageReactionEntity.kind) {
+      switch (eventMessage.content) {
+        case 'received':
+          await _updateConversationMessageAsReceived(eventMessage);
+        case 'read':
+          await _updateConversationMessagesAsRead(eventMessage);
+        default:
+          await _insertConversationReactionsTableData(eventMessage);
+      }
+    }
+  }
+
+  Future<void> _updateConversationMessagesTable(
+    EventMessage eventMessage,
+  ) async {
+    if (eventMessage.kind == PrivateDirectMessageEntity.kind) {
+      final conversationMessage = PrivateDirectMessageEntity.fromEventMessage(eventMessage);
+      final conversationIdByPubkeys = await _lookupConversationByPubkeys(conversationMessage);
+
+      if (conversationIdByPubkeys != null) {
+        // Existing conversation (one-to-one or group)
+        await _insertConversationData(
+          eventMessage: eventMessage,
+          conversationId: conversationIdByPubkeys,
+        );
+      } else {
+        // Existing group conversation (change of participants)
+        final conversationIdBySubject = await _lookupConversationBySubject(conversationMessage);
+
+        if (conversationIdBySubject != null) {
+          await _insertConversationData(
+            eventMessage: eventMessage,
+            conversationId: conversationIdBySubject,
+          );
+        } else if (eventMessage.content.isEmpty) {
+          // New conversation
+          final uuid = const Uuid().v1();
+          await _insertConversationData(
+            eventMessage: eventMessage,
+            conversationId: uuid,
+          );
+        } else {
+          // Invalid message (doesn't belong to any conversation)
+          throw ConversationIsNotFoundException();
+        }
+      }
+    }
+  }
+
+  // Check if there are conversations with the same pubkeys
+  Future<String?> _lookupConversationByPubkeys(
+    PrivateDirectMessageEntity conversationMessage,
+  ) async {
+    final conversationsWithSameParticipants = await (_db.select(_db.conversationMessagesTable)
+          ..where(
+            (table) => table.pubKeys.equals(conversationMessage.allPubkeysMask),
+          )
+          ..limit(1))
+        .get();
+
+    if (conversationsWithSameParticipants.isNotEmpty) {
+      return conversationsWithSameParticipants.first.conversationId;
+    }
+
+    return null;
+  }
+
+  // Check if there are conversations with the same subject but different pubkeys
+  // (this means that amount of participants in conversation was changed)
+  Future<String?> _lookupConversationBySubject(
+    PrivateDirectMessageEntity conversationMessage,
+  ) async {
+    final subject = conversationMessage.data.relatedSubject?.value;
+
+    if (subject != null) {
```

Comment:
It's better to `if (subject == null) return null;`

Reply:
Okay